README.rst
.. image:: https://requires.io/github/Dennis-van-Gils/project-BME280-DS18B20-logger/requirements.svg?branch=master
    :target: https://requires.io/github/Dennis-van-Gils/project-BME280-DS18B20-logger/requirements/?branch=master
    :alt: Requirements Status
.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/psf/black
.. image:: https://img.shields.io/badge/License-MIT-purple.svg
    :target: https://github.com/Dennis-van-Gils/project-BME280-DS18B20-logger/blob/master/LICENSE.txt

BME280 & DS18B20 logger
=======================
*A Physics of Fluids project.*

A temperature, humidity and pressure data logger build from an Adafruit Feather
M4 Express micro-controller board and a BME280 and DS18B20 sensor with a
graphical user interface in Python.

- Github: https://github.com/Dennis-van-Gils/project-BME280-DS18B20-logger

.. image:: https://raw.githubusercontent.com/Dennis-van-Gils/project-BME280-DS18B20-logger/master/images/screenshot.png

Instructions
============
Download the latest release of `BME280 & DS18B20 logger <https://github.com/Dennis-van-Gils/project-BME280-DS18B20-logger/releases/latest>`_
and unpack to a folder onto your drive.

Flashing the firmware
---------------------

Double click the reset button of the Feather while plugged into your PC. This
will mount a drive called `FEATHERBOOT`. Copy
`src_mcu/_build_Feather_M4/CURRENT.UF2 <https://github.com/Dennis-van-Gils/project-BME280-DS18B20-logger/raw/master/src_mcu/_build_Feather_M4/CURRENT.UF2>`_
onto the Featherboot drive. It will restart automatically with the new
firmware.

Running the application
-----------------------

Preferred Python distributions:
    * `Anaconda <https://www.anaconda.com>`_
    * `Miniconda <https://docs.conda.io/en/latest/miniconda.html>`_

Open `Anaconda Prompt` and navigate to the unpacked folder. Run the following to
install the necessary packages: ::

    cd src_python
    pip install -r requirements.txt
    
Now you can run the application: ::

    python main.py

LED status lights
=================

* Solid blue: Booting and setting up
* Solid green: Ready for communication
* Flashing green: Sensor data is being send over USB


CHANGELOG.rst
Changelog
=========

1.0.0 (2020-08-12)
------------------
* Initial release


src_mcu/src/main.cpp
/*******************************************************************************
  BME280 & DS18B20 logger
  Adafruit Feather M4

  BME280:
    Reads out temperature, humidity and pressure over I2C.
    Pins: SDA & SCL
  DS18B20:
    Temperature
    Pins: DI5

  The RGB LED of the Feather M4 will indicate its status:
  * Blue : We're setting up
  * Green: Running okay
  Every read out, the LED will flash brightly in green

  Dennis van Gils
  12-08-2020
*******************************************************************************/

#include <Arduino.h>
#include <DvG_SerialCommand.h>
#include <Adafruit_NeoPixel.h>

// DS18B20
#include <OneWire.h>
#include <DallasTemperature.h>

// BME280
#include <BME280I2C.h>
#include <Wire.h>
#include <SPI.h>

DvG_SerialCommand sc(Serial); // Instantiate serial command listener

Adafruit_NeoPixel neo(1, PIN_NEOPIXEL, NEO_GRB + NEO_KHZ800);
#define NEO_DIM 3  // Brightness level for dim intensity [0 -255]
#define NEO_BRIGHT 8 // Brightness level for bright intensity [0 - 255]

#define PIN_DS18B20 5
OneWire oneWire(PIN_DS18B20);
DallasTemperature ds18(&oneWire);
BME280I2C bme;

float ds18_temp(NAN);     // ['C]
float bme280_temp(NAN);   // ['C]
float bme280_humi(NAN);   // [%]
float bme280_pres(NAN);   // [Pa]

BME280::TempUnit tempUnit(BME280::TempUnit_Celsius);
BME280::PresUnit presUnit(BME280::PresUnit_Pa);

// -----------------------------------------------------------------------------
//    setup
// -----------------------------------------------------------------------------

void setup() {
    neo.begin();
    neo.setPixelColor(0, neo.Color(0, 0, NEO_BRIGHT)); // Blue: We're in setup()
    neo.show();

    Serial.begin(9600);
    ds18.begin();

    // BME280
    Wire.begin();
    while (!bme.begin()) {
        Serial.println("Could not find a valid BME280 sensor, check wiring!");
        delay(1000);
    }

    // Ditch the first reading. The first humidity reading tends to be off.
    bme.read(bme280_pres, bme280_temp, bme280_humi, tempUnit, presUnit);

    neo.setPixelColor(0, neo.Color(0, NEO_DIM, 0)); // Green: All set up
    neo.show();
}

// -----------------------------------------------------------------------------
//    loop
// -----------------------------------------------------------------------------

void loop() {
    char *strCmd; // Incoming serial command string
    uint32_t now;

    if (sc.available()) {
        strCmd = sc.getCmd();

        neo.setPixelColor(0, neo.Color(0, NEO_BRIGHT, 0)); // Green: Flash
        neo.show();

        if (strcmp(strCmd, "id?") == 0) {
            Serial.println("Arduino, BME280 & DS18B20 logger");

        } else {
            now = millis();
            ds18.requestTemperatures();
            ds18_temp = ds18.getTempCByIndex(0);
            bme.read(bme280_pres, bme280_temp, bme280_humi, tempUnit, presUnit);

            Serial.println(
                String(now) +
                '\t' + String(ds18_temp, 1) +
                '\t' + String(bme280_temp, 1) +
                '\t' + String(bme280_humi, 1) +
                '\t' + String(bme280_pres, 0));
        }

        neo.setPixelColor(0, neo.Color(0, NEO_DIM, 0)); // Green: Idle
        neo.show();
    }
}


src_python/DvG_pyqt_controls.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Mishmash of PyQt5 stylesheets and custom controls that I personally use in
many of my projects.
"""
__author__      = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__         = "https://github.com/Dennis-van-Gils/DvG_PyQt_misc"
__date__        = "23-08-2018"
__version__     = "1.0.0"

from PyQt5 import QtWidgets as QtWid

COLOR_RED            = "rgb(255, 0, 0)"
COLOR_YELLOW         = "rgb(255, 255, 0)"
COLOR_INDIAN_RED     = "rgb(205, 92, 92)"
COLOR_SPRING_GREEN_2 = "rgb(0, 238, 118)"
COLOR_BISQUE_5       = "rgb(252, 208, 173)"
COLOR_READ_ONLY      = "rgb(250, 230, 210)"

# Retrieve the standard background color from
# QtGui.QApplication.palette().button().color().name()
# However, at init there is no QApplication instance yet and Python crashes
# hence hard-code here.
COLOR_BG = "rgb(240, 240, 240)"

# Style sheets

SS_TEXTBOX_READ_ONLY = (
        "QLineEdit {"
            "border: 1px solid black}"
        "QLineEdit::read-only {"
            "border: 1px solid gray;"
            "background: " + COLOR_READ_ONLY + "}"
        "QPlainTextEdit {"
            "border: 1px solid black}"
        "QPlainTextEdit[readOnly=\"true\"] {"
            "border: 1px solid gray;"
            "background-color: " + COLOR_READ_ONLY + "}")

SS_TEXTBOX_ERRORS = (
        "QLineEdit {"
            "border: 1px solid gray;"
            "background: " + COLOR_READ_ONLY + "}"
        "QLineEdit::read-only {"
            "border: 2px solid red;"
            "background: yellow;"
            "color: black}"
        "QPlainTextEdit {"
            "border: 1px solid gray;"
            "background-color: " + COLOR_READ_ONLY + "}"
        "QPlainTextEdit[readOnly=\"true\"] {"
            "border: 2px solid red;"
            "background-color: yellow;"
            "color: black}")

SS_GROUP = (
        "QGroupBox {"
            "background-color: " + COLOR_BISQUE_5 + ";"
            "border: 2px solid gray;"
            "border-radius: 5px;"
            "font: bold;"
            "padding: 8 0 0 0px;"
            "margin-top: 2ex}"
        "QGroupBox::title {"
            "subcontrol-origin: margin;"
            "subcontrol-position: top left;"
            "padding: 0 3px}"
        "QGroupBox::flat {"
            "border: 0px;"
            "border-radius: 0 0px;"
            "padding: 0}")

SS_LED = (
        "QPushButton {"
            "background-color: " + COLOR_INDIAN_RED + ";"
            "border-style: inset;"
            "border-width: 1px;"
            "min-height: 30px;"
            "min-width: 30px;"
            "max-width: 30px}"
        "QPushButton::disabled {"
            "border-radius: 15px;"
            "color: black}"
        "QPushButton::checked {"
            "background-color: " + COLOR_SPRING_GREEN_2 + ";"
            "border-style: outset}")

SS_ERROR_LED = (
        "QPushButton {"
            "background-color: " + COLOR_SPRING_GREEN_2 + ";"
            "border: 1px solid gray;"
            "min-height: 30px;"
            "min-width: 30px}"
        "QPushButton::disabled {"
            "color: black}"
        "QPushButton::checked {"
            "font-weight: bold;"
            "background-color: red}")

SS_TINY_ERROR_LED = (
        "QPushButton {"
            "background-color: " + COLOR_BG + ";"
            "border-style: inset;"
            "border-width: 1px;"
            "max-height: 10px;"
            "max-width: 10px;"
            "height: 10px;"
            "width: 10px}"
        "QPushButton::disabled {"
            "border-radius: 5px;"
            "color: black}"
        "QPushButton::checked {"
            "background-color: red;"
            "border-style: outset}")

SS_TINY_LED = (
        "QPushButton {"
            "background-color: " + COLOR_BG + ";"
            "border-style: inset;"
            "border-width: 1px;"
            "max-height: 10px;"
            "max-width: 10px;"
            "height: 10px;"
            "width: 10px}"
        "QPushButton::disabled {"
            "border-radius: 5px;"
            "color: black}"
        "QPushButton::checked {"
            "background-color: " + COLOR_SPRING_GREEN_2 + ";"
            "border-style: outset}")

SS_TEXT_MSGS = (
        "QPlainTextEdit::disabled {"
            "color: black;"
            "background-color: white}")

SS_TITLE = (
        "QLabel {"
            "background-color: " + COLOR_BISQUE_5 + ";"
            "font: bold}")

# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------

def create_Toggle_button_style_sheet(bg_clr=COLOR_BG,
                                     color='black',
                                     checked_bg_clr=COLOR_SPRING_GREEN_2,
                                     checked_color='black',
                                     checked_font_weight='normal'):
    return ("QPushButton {"
                "background-color: " + bg_clr + ";" +
                "color: "            + color + ";" +
                "border-style: outset;"
                "border-width: 2px;"
                "border-radius: 4px;"
                "border-color: gray;"
                "text-align: center;"
                "padding: 1px 1px 1px 1px;}"
            "QPushButton:disabled {"
                "color: black;}"
            "QPushButton:checked {"
                "background-color: " + checked_bg_clr + ";"
                "color: "            + checked_color + ";" +
                "font-weight: "      + checked_font_weight + ";"
                "border-style: inset;}")

def create_LED_indicator():
    button = QtWid.QPushButton("0", checkable=True, enabled=False)
    button.setStyleSheet(SS_LED)
    return button

def create_Relay_button():
    button = QtWid.QPushButton("0", checkable=True)
    button.setStyleSheet(SS_LED)
    return button

def create_Toggle_button(text='', minimumHeight=40):
    SS = ("QPushButton {"
              "background-color: " + COLOR_BG + ";" +
              "color: black;" +
              "border-style: outset;"
              "border-width: 2px;"
              "border-radius: 4px;"
              "border-color: gray;"
              "text-align: center;"
              "padding: 1px 1px 1px 1px;}"
          "QPushButton:disabled {"
              "color: grey;}"
          "QPushButton:checked {"
              "background-color: " + COLOR_SPRING_GREEN_2 + ";"
              "color: black;" +
              "font-weight: normal;"
              "border-style: inset;}")
    button = QtWid.QPushButton(text, checkable=True)
    button.setStyleSheet(SS)

    if minimumHeight is not None:
        button.setMinimumHeight(minimumHeight)

    return button

def create_Toggle_button_2(text='', minimumHeight=40):
    SS = ("QPushButton {"
              "background-color: " + COLOR_BG + ";" +
              "color: black;" +
              "border-style: outset;"
              "border-width: 2px;"
              "border-radius: 4px;"
              "border-color: gray;"
              "text-align: center;"
              "padding: 1px 1px 1px 1px;}"
          "QPushButton:disabled {"
              "color: grey;}"
          "QPushButton:checked {"
              "background-color: " + COLOR_YELLOW + ";"
              "color: black;" +
              "font-weight: bold;"
              "border-color: red;"
              "border-style: inset;}")
    button = QtWid.QPushButton(text, checkable=True)
    button.setStyleSheet(SS)

    if minimumHeight is not None:
        button.setMinimumHeight(minimumHeight)

    return button

def create_Toggle_button_3(text='', minimumHeight=40):
    SS = ("QPushButton {"
              "background-color: " + COLOR_YELLOW + ";" +
              "color: black;" +
              "border-style: outset;"
              "border-width: 2px;"
              "border-radius: 4px;"
              "border-color: red;"
              "text-align: center;"
              "font-weight: bold;"
              "padding: 1px 1px 1px 1px;}"
          "QPushButton:disabled {"
              "color: grey;}"
          "QPushButton:checked {"
              "background-color: " + COLOR_SPRING_GREEN_2 + ";"
              "color: black;" +
              "border-color: gray;"
              "border-style: inset;"
              "font-weight: normal;}")
    button = QtWid.QPushButton(text, checkable=True)
    button.setStyleSheet(SS)

    if minimumHeight is not None:
        button.setMinimumHeight(minimumHeight)

    return button

def create_tiny_error_LED(text=''):
    button = QtWid.QPushButton(text='', checkable=True, enabled=False)
    button.setStyleSheet(SS_TINY_ERROR_LED)
    return button

def create_tiny_LED(text=''):
    button = QtWid.QPushButton(text='', checkable=True, enabled=False)
    button.setStyleSheet(SS_TINY_LED)
    return button

def create_error_LED(text=''):
    button = QtWid.QPushButton(text='', checkable=True, enabled=False)
    button.setStyleSheet(SS_ERROR_LED)
    return button


src_python/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""BME280 & DS18B20 logger
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-BME280-DS18B20-logger"
__date__ = "12-08-2020"
__version__ = "1.0"
# pylint: disable=bare-except, broad-except

import os
import sys
import time

import numpy as np
import psutil

from PyQt5 import QtCore, QtGui
from PyQt5 import QtWidgets as QtWid
from PyQt5.QtCore import QDateTime
import pyqtgraph as pg

from dvg_debug_functions import tprint, dprint, print_fancy_traceback as pft
from dvg_pyqt_controls import (
    create_Toggle_button,
    SS_TEXTBOX_READ_ONLY,
    SS_GROUP,
)
from dvg_pyqt_filelogger import FileLogger
from dvg_pyqtgraph_threadsafe import (
    HistoryChartCurve,
    LegendSelect,
    PlotManager,
)

from dvg_devices.Arduino_protocol_serial import Arduino
from dvg_qdeviceio import QDeviceIO


TRY_USING_OPENGL = True
if TRY_USING_OPENGL:
    try:
        import OpenGL.GL as gl  # pylint: disable=unused-import
    except:
        print("OpenGL acceleration: Disabled")
        print("To install: `conda install pyopengl` or `pip install pyopengl`")
    else:
        print("OpenGL acceleration: Enabled")
        pg.setConfigOptions(useOpenGL=True)
        pg.setConfigOptions(antialias=True)
        pg.setConfigOptions(enableExperimental=True)

# Global pyqtgraph configuration
# pg.setConfigOptions(leftButtonPan=False)
pg.setConfigOption("foreground", "#EEE")

# Constants
# fmt: off
DAQ_INTERVAL_MS    = 1000  # [ms]
CHART_INTERVAL_MS  = 500   # [ms]
CHART_HISTORY_TIME = 3600  # [s]
# fmt: on

# Show debug info in terminal? Warning: Slow! Do not leave on unintentionally.
DEBUG = False


def get_current_date_time():
    cur_date_time = QDateTime.currentDateTime()
    return (
        cur_date_time.toString("dd-MM-yyyy"),  # Date
        cur_date_time.toString("HH:mm:ss"),  # Time
        cur_date_time.toString("yyMMdd_HHmmss"),  # Reverse notation date-time
    )


# ------------------------------------------------------------------------------
#   Arduino state
# ------------------------------------------------------------------------------


class State(object):
    """Reflects the actual readings, parsed into separate variables, of the
    Arduino. There should only be one instance of the State class.
    """

    def __init__(self):
        self.time = np.nan  # [s]
        self.ds_temp = np.nan  # ['C]
        self.bme_temp = np.nan  # ['C]
        self.bme_humi = np.nan  # [%]
        self.bme_pres = np.nan  # [bar]


state = State()

# ------------------------------------------------------------------------------
#   MainWindow
# ------------------------------------------------------------------------------


class MainWindow(QtWid.QWidget):
    def __init__(self, parent=None, **kwargs):
        super().__init__(parent, **kwargs)

        self.setWindowTitle("BME280 & DS18B20 logger")
        self.setGeometry(350, 50, 960, 800)
        self.setStyleSheet(SS_TEXTBOX_READ_ONLY + SS_GROUP)

        # -------------------------
        #   Top frame
        # -------------------------

        # Left box
        self.qlbl_update_counter = QtWid.QLabel("0")
        self.qlbl_DAQ_rate = QtWid.QLabel("DAQ: nan Hz")
        self.qlbl_DAQ_rate.setStyleSheet("QLabel {min-width: 7em}")

        vbox_left = QtWid.QVBoxLayout()
        vbox_left.addWidget(self.qlbl_update_counter, stretch=0)
        vbox_left.addStretch(1)
        vbox_left.addWidget(self.qlbl_DAQ_rate, stretch=0)

        # Middle box
        self.qlbl_title = QtWid.QLabel(
            "BME280 & DS18B20 logger",
            font=QtGui.QFont("Palatino", 14, weight=QtGui.QFont.Bold),
        )
        self.qlbl_title.setAlignment(QtCore.Qt.AlignCenter)
        self.qlbl_cur_date_time = QtWid.QLabel("00-00-0000    00:00:00")
        self.qlbl_cur_date_time.setAlignment(QtCore.Qt.AlignCenter)
        self.qpbt_record = create_Toggle_button(
            "Click to start recording to file"
        )
        # fmt: off
        self.qpbt_record.clicked.connect(lambda state: log.record(state)) # pylint: disable=unnecessary-lambda
        # fmt: on

        vbox_middle = QtWid.QVBoxLayout()
        vbox_middle.addWidget(self.qlbl_title)
        vbox_middle.addWidget(self.qlbl_cur_date_time)
        vbox_middle.addWidget(self.qpbt_record)

        # Right box
        self.qpbt_exit = QtWid.QPushButton("Exit")
        self.qpbt_exit.clicked.connect(self.close)
        self.qpbt_exit.setMinimumHeight(30)
        self.qlbl_recording_time = QtWid.QLabel(alignment=QtCore.Qt.AlignRight)

        vbox_right = QtWid.QVBoxLayout()
        vbox_right.addWidget(self.qpbt_exit, stretch=0)
        vbox_right.addStretch(1)
        vbox_right.addWidget(self.qlbl_recording_time, stretch=0)

        # Round up top frame
        hbox_top = QtWid.QHBoxLayout()
        hbox_top.addLayout(vbox_left, stretch=0)
        hbox_top.addStretch(1)
        hbox_top.addLayout(vbox_middle, stretch=0)
        hbox_top.addStretch(1)
        hbox_top.addLayout(vbox_right, stretch=0)

        # -------------------------
        #   Bottom frame
        # -------------------------

        #  Charts
        # -------------------------

        self.gw = pg.GraphicsLayoutWidget()

        # Plot: Temperatures
        p = {"color": "#EEE", "font-size": "10pt"}
        self.pi_temp = self.gw.addPlot(row=0, col=0)
        self.pi_temp.setLabel("left", text="temperature (°C)", **p)

        # Plot: Humidity
        self.pi_humi = self.gw.addPlot(row=1, col=0)
        self.pi_humi.setLabel("left", text="humidity (%)", **p)

        # Plot: Pressure
        self.pi_pres = self.gw.addPlot(row=2, col=0)
        self.pi_pres.setLabel("left", text="pressure (mbar)", **p)

        self.plots = [self.pi_temp, self.pi_humi, self.pi_pres]
        for plot in self.plots:
            plot.setClipToView(True)
            plot.showGrid(x=1, y=1)
            plot.setLabel("bottom", text="history (s)", **p)
            plot.setMenuEnabled(True)
            plot.enableAutoRange(axis=pg.ViewBox.XAxis, enable=False)
            plot.enableAutoRange(axis=pg.ViewBox.YAxis, enable=True)
            plot.setAutoVisible(y=True)
            plot.setRange(xRange=[-CHART_HISTORY_TIME, 0])

        # Curves
        capacity = round(CHART_HISTORY_TIME * 1e3 / DAQ_INTERVAL_MS)
        PEN_01 = pg.mkPen(color=[255, 255, 0], width=3)
        PEN_02 = pg.mkPen(color=[0, 255, 255], width=3)

        self.tscurve_ds_temp = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_temp.plot(pen=PEN_01, name="temp_DS"),
        )
        self.tscurve_bme_temp = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_temp.plot(pen=PEN_02, name="temp_BME"),
        )
        self.tscurve_bme_humi = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_humi.plot(pen=PEN_02, name="humi_BME"),
        )
        self.tscurve_bme_pres = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_pres.plot(pen=PEN_02, name="pres_BME"),
        )

        self.tscurves = [
            self.tscurve_ds_temp,
            self.tscurve_bme_temp,
            self.tscurve_bme_humi,
            self.tscurve_bme_pres,
        ]

        #  Group `Readings`
        # -------------------------

        legend = LegendSelect(
            linked_curves=self.tscurves, hide_toggle_button=True
        )

        p = {
            "readOnly": True,
            "alignment": QtCore.Qt.AlignRight,
            "maximumWidth": 54,
        }
        self.qlin_ds_temp = QtWid.QLineEdit(**p)
        self.qlin_bme_temp = QtWid.QLineEdit(**p)
        self.qlin_bme_humi = QtWid.QLineEdit(**p)
        self.qlin_bme_pres = QtWid.QLineEdit(**p)

        # fmt: off
        legend.grid.setHorizontalSpacing(6)
        legend.grid.addWidget(self.qlin_ds_temp       , 0, 2)
        legend.grid.addWidget(QtWid.QLabel("± 0.5 °C"), 0, 3)
        legend.grid.addWidget(self.qlin_bme_temp      , 1, 2)
        legend.grid.addWidget(QtWid.QLabel("± 0.5 °C"), 1, 3)
        legend.grid.addWidget(self.qlin_bme_humi      , 2, 2)
        legend.grid.addWidget(QtWid.QLabel("± 3 %")   , 2, 3)
        legend.grid.addWidget(self.qlin_bme_pres      , 3, 2)
        legend.grid.addWidget(QtWid.QLabel("± 1 mbar"), 3, 3)
        # fmt: on

        qgrp_readings = QtWid.QGroupBox("Readings")
        qgrp_readings.setLayout(legend.grid)

        #  Group 'Log comments'
        # -------------------------

        self.qtxt_comments = QtWid.QTextEdit()
        grid = QtWid.QGridLayout()
        grid.addWidget(self.qtxt_comments, 0, 0)

        qgrp_comments = QtWid.QGroupBox("Log comments")
        qgrp_comments.setLayout(grid)

        #  Group 'Charts'
        # -------------------------

        self.plot_manager = PlotManager(parent=self)
        self.plot_manager.add_autorange_buttons(linked_plots=self.plots)
        self.plot_manager.add_preset_buttons(
            linked_plots=self.plots,
            linked_curves=self.tscurves,
            presets=[
                {
                    "button_label": "00:30",
                    "x_axis_label": "history (sec)",
                    "x_axis_divisor": 1,
                    "x_axis_range": (-30, 0),
                },
                {
                    "button_label": "01:00",
                    "x_axis_label": "history (sec)",
                    "x_axis_divisor": 1,
                    "x_axis_range": (-60, 0),
                },
                {
                    "button_label": "10:00",
                    "x_axis_label": "history (min)",
                    "x_axis_divisor": 60,
                    "x_axis_range": (-10, 0),
                },
                {
                    "button_label": "30:00",
                    "x_axis_label": "history (min)",
                    "x_axis_divisor": 60,
                    "x_axis_range": (-30, 0),
                },
                {
                    "button_label": "60:00",
                    "x_axis_label": "history (min)",
                    "x_axis_divisor": 60,
                    "x_axis_range": (-60, 0),
                },
            ],
        )
        self.plot_manager.add_clear_button(linked_curves=self.tscurves)
        self.plot_manager.perform_preset(1)

        qgrp_chart = QtWid.QGroupBox("Charts")
        qgrp_chart.setLayout(self.plot_manager.grid)

        vbox = QtWid.QVBoxLayout()
        vbox.addWidget(qgrp_readings)
        vbox.addWidget(qgrp_comments)
        vbox.addWidget(qgrp_chart, alignment=QtCore.Qt.AlignLeft)
        vbox.addStretch()

        # Round up bottom frame
        hbox_bot = QtWid.QHBoxLayout()
        hbox_bot.addWidget(self.gw, 1)
        hbox_bot.addLayout(vbox, 0)

        # -------------------------
        #   Round up full window
        # -------------------------

        vbox = QtWid.QVBoxLayout(self)
        vbox.addLayout(hbox_top, stretch=0)
        vbox.addSpacerItem(QtWid.QSpacerItem(0, 10))
        vbox.addLayout(hbox_bot, stretch=1)

    # --------------------------------------------------------------------------
    #   Handle controls
    # --------------------------------------------------------------------------

    @QtCore.pyqtSlot()
    def update_GUI(self):
        str_cur_date, str_cur_time, _ = get_current_date_time()
        self.qlbl_cur_date_time.setText(
            "%s    %s" % (str_cur_date, str_cur_time)
        )
        self.qlbl_update_counter.setText("%i" % qdev_ard.update_counter_DAQ)
        self.qlbl_DAQ_rate.setText(
            "DAQ: %.1f Hz" % qdev_ard.obtained_DAQ_rate_Hz
        )
        if log.is_recording():
            self.qlbl_recording_time.setText(log.pretty_elapsed())

        self.qlin_ds_temp.setText("%.1f" % state.ds_temp)
        self.qlin_bme_temp.setText("%.1f" % state.bme_temp)
        self.qlin_bme_humi.setText("%.1f" % state.bme_humi)
        self.qlin_bme_pres.setText("%.1f" % state.bme_pres)

    @QtCore.pyqtSlot()
    def update_chart(self):
        if DEBUG:
            tprint("update_chart")

        for tscurve in self.tscurves:
            tscurve.update()


# ------------------------------------------------------------------------------
#   Program termination routines
# ------------------------------------------------------------------------------


def stop_running():
    app.processEvents()
    qdev_ard.quit()
    log.close()

    print("Stopping timers................ ", end="")
    timer_GUI.stop()
    timer_charts.stop()
    print("done.")


@QtCore.pyqtSlot()
def notify_connection_lost():
    stop_running()

    window.qlbl_title.setText("! ! !    LOST CONNECTION    ! ! !")
    str_cur_date, str_cur_time, _ = get_current_date_time()
    str_msg = "%s %s\nLost connection to Arduino." % (
        str_cur_date,
        str_cur_time,
    )
    print("\nCRITICAL ERROR @ %s" % str_msg)
    reply = QtWid.QMessageBox.warning(
        window, "CRITICAL ERROR", str_msg, QtWid.QMessageBox.Ok
    )

    if reply == QtWid.QMessageBox.Ok:
        pass  # Leave the GUI open for read-only inspection by the user


@QtCore.pyqtSlot()
def about_to_quit():
    print("\nAbout to quit")
    stop_running()
    ard.close()


# ------------------------------------------------------------------------------
#   Your Arduino update function
# ------------------------------------------------------------------------------


def DAQ_function():
    # Date-time keeping
    str_cur_date, str_cur_time, str_cur_datetime = get_current_date_time()

    # Query the Arduino for its state
    success, tmp_state = ard.query_ascii_values("?", delimiter="\t")
    if not (success):
        dprint(
            "'%s' reports IOError @ %s %s"
            % (ard.name, str_cur_date, str_cur_time)
        )
        return False

    # Parse readings into separate state variables
    try:
        (
            state.time,
            state.ds_temp,
            state.bme_temp,
            state.bme_humi,
            state.bme_pres,
        ) = tmp_state
        state.time /= 1000  # Arduino time, [msec] to [s]
        state.bme_pres /= 100  # [Pa] to [mbar]
    except Exception as err:
        pft(err, 3)
        dprint(
            "'%s' reports IOError @ %s %s"
            % (ard.name, str_cur_date, str_cur_time)
        )
        return False

    # We will use PC time instead
    state.time = time.perf_counter()

    # Add readings to chart histories
    window.tscurve_ds_temp.appendData(state.time, state.ds_temp)
    window.tscurve_bme_temp.appendData(state.time, state.bme_temp)
    window.tscurve_bme_humi.appendData(state.time, state.bme_humi)
    window.tscurve_bme_pres.appendData(state.time, state.bme_pres)

    # Logging to file
    log.update(filepath=str_cur_datetime + ".txt", mode="w")

    # Return success
    return True


def write_header_to_log():
    log.write("[HEADER]\n")
    log.write(window.qtxt_comments.toPlainText())
    log.write("\n\n[DATA]\n")
    log.write("time\ttemp_DS\ttemp_BME\thumi_BME\tpres_BME\n")
    log.write("[s]\t[±0.5 °C]\t[±0.5 °C]\t[±3 pct]\t[±1 mbar]\n")


def write_data_to_log():
    log.write(
        "%.1f\t%.1f\t%.1f\t%.1f\t%.1f\n"
        % (
            log.elapsed(),
            state.ds_temp,
            state.bme_temp,
            state.bme_humi,
            state.bme_pres,
        )
    )


# ------------------------------------------------------------------------------
#   Main
# ------------------------------------------------------------------------------

if __name__ == "__main__":
    # Set priority of this process to maximum in the operating system
    print("PID: %s\n" % os.getpid())
    try:
        proc = psutil.Process(os.getpid())
        if os.name == "nt":
            proc.nice(psutil.REALTIME_PRIORITY_CLASS)  # Windows
        else:
            proc.nice(-20)  # Other
    except:
        print("Warning: Could not set process to maximum priority.\n")

    # --------------------------------------------------------------------------
    #   Connect to Arduino
    # --------------------------------------------------------------------------

    ard = Arduino(name="Ard", connect_to_specific_ID="BME280 & DS18B20 logger")
    ard.serial_settings["baudrate"] = 115200
    ard.auto_connect()

    if not (ard.is_alive):
        print("\nCheck connection and try resetting the Arduino.")
        print("Exiting...\n")
        sys.exit(0)

    # --------------------------------------------------------------------------
    #   Create application and main window
    # --------------------------------------------------------------------------
    QtCore.QThread.currentThread().setObjectName("MAIN")  # For DEBUG info

    app = QtWid.QApplication(sys.argv)
    app.aboutToQuit.connect(about_to_quit)

    window = MainWindow()

    # --------------------------------------------------------------------------
    #   File logger
    # --------------------------------------------------------------------------

    log = FileLogger(
        write_header_function=write_header_to_log,
        write_data_function=write_data_to_log,
    )
    log.signal_recording_started.connect(
        lambda filepath: window.qpbt_record.setText(
            "Recording to file: %s" % filepath
        )
    )
    log.signal_recording_stopped.connect(
        lambda: window.qpbt_record.setText("Click to start recording to file")
    )

    # --------------------------------------------------------------------------
    #   Set up multithreaded communication with the Arduino
    # --------------------------------------------------------------------------

    # Create QDeviceIO
    qdev_ard = QDeviceIO(ard)

    # Create workers
    # fmt: off
    qdev_ard.create_worker_DAQ(
        DAQ_function             = DAQ_function,
        DAQ_interval_ms          = DAQ_INTERVAL_MS,
        critical_not_alive_count = 3,
        debug                    = DEBUG,
    )
    # fmt: on

    # Connect signals to slots
    qdev_ard.signal_DAQ_updated.connect(window.update_GUI)
    qdev_ard.signal_connection_lost.connect(notify_connection_lost)

    # Start workers
    qdev_ard.start(DAQ_priority=QtCore.QThread.TimeCriticalPriority)

    # --------------------------------------------------------------------------
    #   Timers
    # --------------------------------------------------------------------------

    timer_GUI = QtCore.QTimer()
    timer_GUI.timeout.connect(window.update_GUI)
    timer_GUI.start(100)

    timer_charts = QtCore.QTimer()
    timer_charts.timeout.connect(window.update_chart)
    timer_charts.start(CHART_INTERVAL_MS)

    # --------------------------------------------------------------------------
    #   Start the main GUI event loop
    # --------------------------------------------------------------------------

    window.show()
    sys.exit(app.exec_())
